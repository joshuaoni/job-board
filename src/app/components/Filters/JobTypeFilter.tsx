"use client";

import { useState, useEffect, useRef } from "react";
import Image from "next/image";
import { FilterProps, JOB_TYPE_OPTIONS } from "@/app/types/filters";
import { animations } from "@/app/utils/animations";

export const JobTypeFilter = ({ value, onChange }: FilterProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsOpen(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  const selectedOption = JOB_TYPE_OPTIONS.find(
    (option) => option.key === value
  );

  return (
    <div ref={dropdownRef} className="relative w-[218px]">
      <button
        type="button"
        className={`w-full h-[44px] rounded-lg px-4 bg-[#EBEBEB] text-black flex justify-between items-center hover:bg-[#E5E5E5] transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-gray-300 ${
          isOpen ? "ring-2 ring-gray-300" : ""
        }`}
        onClick={() => setIsOpen(!isOpen)}
        aria-haspopup="listbox"
        aria-expanded={isOpen}
        aria-label="Select job type"
      >
        <span>{selectedOption?.label || "Select Job Type"}</span>
        <Image
          src="/arrow-down.png"
          alt=""
          width={21}
          height={21}
          className={`transform transition-transform duration-200 ${
            isOpen ? "rotate-180" : ""
          }`}
          priority={false}
          aria-hidden="true"
        />
      </button>

      {isOpen && (
        <ul
          className={`absolute top-[48px] left-0 w-full bg-white text-black shadow-lg rounded-lg z-10 py-1 ${animations.scaleIn} origin-top`}
          role="listbox"
        >
          {JOB_TYPE_OPTIONS.map(({ key, label }) => (
            <li key={key} role="option" aria-selected={value === key}>
              <button
                className="w-full px-4 py-2 text-left hover:bg-gray-100 transition-colors focus:outline-none focus:bg-gray-100"
                onClick={() => {
                  onChange(key);
                  setIsOpen(false);
                }}
              >
                {label}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};
