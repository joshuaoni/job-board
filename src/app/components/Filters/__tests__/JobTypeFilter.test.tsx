import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { JobTypeFilter } from "../JobTypeFilter";

// Mock next/navigation
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
  useSearchParams: () => ({
    get: () => null,
    toString: () => "",
  }),
}));

describe("JobTypeFilter", () => {
  const mockOnChange = jest.fn();

  beforeEach(() => {
    mockOnChange.mockClear();
  });

  it("should render with default state", () => {
    render(<JobTypeFilter value="" onChange={mockOnChange} />);

    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByText("Select Job Type")).toBeInTheDocument();
  });

  it("should show all job type options when clicked", () => {
    render(<JobTypeFilter value="" onChange={mockOnChange} />);

    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(screen.getByRole("option", { name: "Hybrid" })).toBeInTheDocument();
    expect(
      screen.getByRole("option", { name: "Full time" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("option", { name: "Part time" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("option", { name: "Internship" })
    ).toBeInTheDocument();
  });

  it("should call onChange when an option is selected", () => {
    render(<JobTypeFilter value="" onChange={mockOnChange} />);

    const button = screen.getByRole("button");
    fireEvent.click(button);

    const fullTimeOption = screen.getByRole("option", { name: "Full time" });
    const fullTimeButton = fullTimeOption.querySelector("button");
    fireEvent.click(fullTimeButton!);

    expect(mockOnChange).toHaveBeenCalledWith("full_time");
  });

  it("should show selected option", () => {
    render(<JobTypeFilter value="part_time" onChange={mockOnChange} />);

    expect(screen.getByRole("button")).toHaveTextContent("Part time");
  });

  it("should be accessible", () => {
    render(<JobTypeFilter value="" onChange={mockOnChange} />);

    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("aria-label", "Select job type");
    expect(button).toHaveAttribute("aria-expanded", "false");

    fireEvent.click(button);
    expect(button).toHaveAttribute("aria-expanded", "true");
    expect(screen.getByRole("listbox")).toBeInTheDocument();
  });

  it("should close dropdown when clicking outside", () => {
    render(<JobTypeFilter value="" onChange={mockOnChange} />);

    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(screen.getByRole("listbox")).toBeInTheDocument();

    fireEvent.mouseDown(document.body);
    expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
  });
});
