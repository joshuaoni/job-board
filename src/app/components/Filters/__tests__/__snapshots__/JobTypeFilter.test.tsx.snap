// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobTypeFilter should match snapshot in default state 1`] = `
<div>
  <div
    class="relative w-[218px]"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-label="Select job type"
      class="w-full h-[44px] rounded-lg px-4 bg-[#EBEBEB] text-black flex justify-between items-center hover:bg-[#E5E5E5] transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-gray-300 "
      type="button"
    >
      <span>
        Select Job Type
      </span>
      <img
        alt=""
        aria-hidden="true"
        class="transform transition-transform duration-200 "
        data-nimg="1"
        decoding="async"
        height="21"
        loading="lazy"
        src="/_next/image?url=%2Farrow-down.png&w=48&q=75"
        srcset="/_next/image?url=%2Farrow-down.png&w=32&q=75 1x, /_next/image?url=%2Farrow-down.png&w=48&q=75 2x"
        style="color: transparent;"
        width="21"
      />
    </button>
  </div>
</div>
`;

exports[`JobTypeFilter should match snapshot with dropdown open 1`] = `
<div>
  <div
    class="relative w-[218px]"
  >
    <button
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-label="Select job type"
      class="w-full h-[44px] rounded-lg px-4 bg-[#EBEBEB] text-black flex justify-between items-center hover:bg-[#E5E5E5] transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-gray-300 ring-2 ring-gray-300"
      type="button"
    >
      <span>
        Select Job Type
      </span>
      <img
        alt=""
        aria-hidden="true"
        class="transform transition-transform duration-200 rotate-180"
        data-nimg="1"
        decoding="async"
        height="21"
        loading="lazy"
        src="/_next/image?url=%2Farrow-down.png&w=48&q=75"
        srcset="/_next/image?url=%2Farrow-down.png&w=32&q=75 1x, /_next/image?url=%2Farrow-down.png&w=48&q=75 2x"
        style="color: transparent;"
        width="21"
      />
    </button>
    <ul
      class="absolute top-[48px] left-0 w-full bg-white text-black shadow-lg rounded-lg z-10 py-1 animate-[scaleIn_0.2s_ease-out] origin-top"
      role="listbox"
    >
      <li
        aria-selected="false"
        role="option"
      >
        <button
          class="w-full px-4 py-2 text-left hover:bg-gray-100 transition-colors focus:outline-none focus:bg-gray-100"
        >
          Hybrid
        </button>
      </li>
      <li
        aria-selected="false"
        role="option"
      >
        <button
          class="w-full px-4 py-2 text-left hover:bg-gray-100 transition-colors focus:outline-none focus:bg-gray-100"
        >
          Full time
        </button>
      </li>
      <li
        aria-selected="false"
        role="option"
      >
        <button
          class="w-full px-4 py-2 text-left hover:bg-gray-100 transition-colors focus:outline-none focus:bg-gray-100"
        >
          Part time
        </button>
      </li>
      <li
        aria-selected="false"
        role="option"
      >
        <button
          class="w-full px-4 py-2 text-left hover:bg-gray-100 transition-colors focus:outline-none focus:bg-gray-100"
        >
          Internship
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`JobTypeFilter should match snapshot with selected value 1`] = `
<div>
  <div
    class="relative w-[218px]"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-label="Select job type"
      class="w-full h-[44px] rounded-lg px-4 bg-[#EBEBEB] text-black flex justify-between items-center hover:bg-[#E5E5E5] transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-gray-300 "
      type="button"
    >
      <span>
        Full time
      </span>
      <img
        alt=""
        aria-hidden="true"
        class="transform transition-transform duration-200 "
        data-nimg="1"
        decoding="async"
        height="21"
        loading="lazy"
        src="/_next/image?url=%2Farrow-down.png&w=48&q=75"
        srcset="/_next/image?url=%2Farrow-down.png&w=32&q=75 1x, /_next/image?url=%2Farrow-down.png&w=48&q=75 2x"
        style="color: transparent;"
        width="21"
      />
    </button>
  </div>
</div>
`;
